src/java.base/share/classes/java/lang/reflect/code/type/JavaType.java

```diff
@@ -125,44 +122,29 @@ default Class<?> resolve(MethodHandles.Lookup l) throws ReflectiveOperationExcep
 
     // Factories
 
-    static JavaType type(Class<?> c) {
-        if (c.isPrimitive()) {
-            return new PrimitiveType(c.getName());
-        } else if (c.isArray()) {
-            return array(type(c.getComponentType()));
-        } else {
-            return new ClassType(c.getName());
-        }
-    }
-
-    static JavaType type(Class<?> c, Class<?>... typeArguments) {
-        return type(c, List.of(typeArguments));
-    }
-
-    static JavaType type(Class<?> c, List<Class<?>> typeArguments) {
-        if (c.isPrimitive()) {
-            throw new IllegalArgumentException("Cannot parameterize a primitive type");
-        } else if (c.isArray()) {
-            return array(type(c.getComponentType(), typeArguments));
-        } else {
-            return new ClassType(c.getName(),
-                    typeArguments.stream().map(JavaType::type).toList());
-        }
-    }
-
-    static JavaType ofNominalDescriptor(ClassDesc d) {
-        return ofNominalDescriptorStringInternal(d.descriptorString(), 0);
+    static JavaType type(Type t) {
+        return switch (t) {
+            case Class<?> c -> type(c.describeConstable().get());
+            case java.lang.reflect.WildcardType wt -> wt.getLowerBounds().length == 0 ?
+                    wildcard(BoundKind.EXTENDS, type(wt.getUpperBounds()[0])) :
+                    wildcard(BoundKind.SUPER, type(wt.getLowerBounds()[0]));
+            case TypeVariable<?> tv -> typeVarRef(tv.getName(), owner(tv.getGenericDeclaration()), type(tv.getBounds()[0]));
+            case GenericArrayType at -> array(type(at.getGenericComponentType()));
+            default -> throw new InternalError();
+        };
     }
 
-    static JavaType ofNominalDescriptorString(String d) {
-        return ofNominalDescriptor(ClassDesc.ofDescriptor(d));
+    private static TypeVarRef.Owner owner(GenericDeclaration genDecl) {
+        return switch (genDecl) {
+            case Executable e -> MethodRef.method(e);
+            case Class<?> t -> (ClassType)type(t);
+            default -> throw new InternalError();
+        };
     }
 
-    private static JavaType ofNominalDescriptorStringInternal(String descriptor, int i) {
-        if (descriptor.charAt(i) == '[') {
-            return new ArrayType(ofNominalDescriptorStringInternal(descriptor, i + 1));
-        } else {
-            return switch (descriptor.charAt(i)) {
+    static JavaType type(ClassDesc d) {
```

Comment:
I've changed name to this factory (used to be `ofNominalDescriptor`). It seemed better to have a symmetry between factories taking a `j.l.r.Type` and factories taking a `ClassDesc`. But if the old name was preferred we can keep that.